Implement initial fruit shop application and comprehensive tests

Mwxyz2022 opened the pull request:
- added initial source code for the fruit shop application
- developed unit tests from scratch for all core modules•created tests for data access operations (dao)
- created tests for business logic services
- created tests for file reading and writing utilities
- created tests for data conversion processes
- created tests for individual transaction operation strategies
- created tests for data model validation and behavior
- created tests for report generation logic
- ensured test coverage for primary use cases and error scenarios
- utilized junit 5 for writing and running tests
- applied consistent code formatting and style for tests

Mwxyz2022 committed:
Implement initial fruit shop application and comprehensive tests

- added initial source code for the fruit shop application
- developed unit tests from scratch for all core modules•created tests for data access operations (dao)
- created tests for business logic services
- created tests for file reading and writing utilities
- created tests for data conversion processes
- created tests for individual transaction operation strategies
- created tests for data model validation and behavior
- created tests for report generation logic
- ensured test coverage for primary use cases and error scenarios
- utilized junit 5 for writing and running tests
- applied consistent code formatting and style for tests

liliia-ponomarenko commented on src/test/java/core/basesyntax/converter/DataConverterImplTest.java:
```diff
@@ -0,0 +1,126 @@
+        FruitTransaction tx4 = transactions.get(3);
+        assertEquals(FruitTransaction.Operation.RETURN, tx4.getOperation());
+        assertEquals("apple", tx4.getFruit());
+        assertEquals(10, tx4.getQuantity());
```
Can you compare FruitTransaction objects or a collection of the objects?

liliia-ponomarenko commented on src/test/java/core/basesyntax/converter/DataConverterImplTest.java:
```diff
@@ -0,0 +1,126 @@
+        RuntimeException exception = assertThrows(RuntimeException.class,
+                () -> dataConverter.convertToTransaction(input));
+        assertEquals("Invalid quantity format in line: b,banana,abc", exception.getMessage());
+        assertInstanceOf(NumberFormatException.class, exception.getCause());
```
```suggestion
       assertThrows(RuntimeException.class, exception.getCause());
```

Elena-Bruyako replied:
fix comment

liliia-ponomarenko commented on src/main/java/core/basesyntax/dao/FruitDaoImpl.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Map;
+
+public class FruitDaoImpl implements FruitDao {
+    private final Map<String, Integer> fruitStorage = new HashMap<>();
```
The map should be static and imitate real DB

liliia-ponomarenko commented on src/test/java/core/basesyntax/dao/FruitDaoImplTest.java:
```diff
@@ -0,0 +1,121 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class FruitDaoImplTest {
```
Clear storage after testing

liliia-ponomarenko commented on src/test/java/core/basesyntax/io/FileReaderImplTest.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Test
+    void read_existingFileWithContent_returnsCorrectLines() throws IOException {
+        Path testFile = tempDir.resolve("test_input.txt");
```
Move the path to the constant fields

liliia-ponomarenko commented on src/test/java/core/basesyntax/io/FileReaderImplTest.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Test
+    void read_emptyFile_returnsEmptyList() throws IOException {
+        Path testFile = tempDir.resolve("empty_input.txt");
```
the same for other test cases

liliia-ponomarenko commented on src/test/java/core/basesyntax/io/FileWriterImplTest.java:
```diff
@@ -0,0 +1,99 @@
+        Path outputFile = tempDir.resolve("null_report.txt");
+        String filePath = outputFile.toString();
+
+        assertThrows(NullPointerException.class,
```
don't test system exception

liliia-ponomarenko commented on src/test/java/core/basesyntax/io/FileWriterImplTest.java:
```diff
@@ -0,0 +1,99 @@
+    @Test
+    void write_emptyReport_createsEmptyFile() throws IOException {
+        String reportContent = "";
+        Path outputFile = tempDir.resolve("empty_output.csv");
```
Use the test resources directory for the testing files

liliia-ponomarenko commented on src/test/java/core/basesyntax/model/FruitTransactionOperationTest.java:
```diff
@@ -0,0 +1,82 @@
+    @Test
+    void getCode_returnOperation_returnsCorrectCode() {
+        assertEquals("r", FruitTransaction.Operation.RETURN.getCode());
+    }
```
Don't test getters

liliia-ponomarenko commented on src/test/java/core/basesyntax/report/ReportGeneratorImplTest.java:
```diff
@@ -0,0 +1,113 @@
+                    }
+                })
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
```
Don't complicate your test cases. Remove this method

liliia-ponomarenko commented on src/test/java/core/basesyntax/report/ReportGeneratorImplTest.java:
```diff
@@ -0,0 +1,113 @@
+
+        final Map<String, Integer> actualFruitsFromReport = parseReportToMap(actualReport);
+
+        assertEquals(fruitDao.getAll(), actualFruitsFromReport,
+                "Parsed report content does not match the data in DAO for multiple fruits.");
```
just compare report content

liliia-ponomarenko reviewed (changes requested):
Good job! Let's improve your solution ;)

Mwxyz2022 committed:
Improve Testability and Data Handling:

- made fruitStorage in FruitDaoImpl static to create a single, application-wide data source.
- added a new clear() method to the FruitDao interface to allow resetting the data storage.
- implemented the clear() method in FruitDaoImpl to clear the static fruitStorage map.
- added @AfterEach teardown methods to all test classes that use FruitDao to ensure test isolation.
- implemented equals() and hashCode() in the FruitTransaction model to enable proper object comparison.
- refactored DataConverterImplTest to directly compare lists of FruitTransaction objects, making the test more concise.
- refactored FileReaderImplTest to use static files from the src/test/resources directory instead of creating temporary files.
- added test_input.csv and empty_input.csv files to src/test/resources for consistent test data.
- removed the dynamic test file creation logic using @TempDir from FileReaderImplTest.
- improved FileWriterImpl by adding a check to prevent writing null report content.
- updated FileWriterImplTest to expect IllegalArgumentException for null content and use constants for test data.
- simplified ReportGeneratorImplTest by removing a complex report parsing helper method.
- replaced detailed map-based assertions in ReportGeneratorImplTest with simpler string-based checks (contains, startsWith).
- added cleanup logic using @AfterEach to ReturnOperationTest and SupplyOperationTest to maintain test independence.
- added cleanup logic using @AfterEach to ShopServiceImplTest for better test isolation.
- removed redundant getter tests from FruitTransactionOperationTest to make the test suite cleaner.

Elena-Bruyako commented on src/test/java/core/basesyntax/report/ReportGeneratorImplTest.java:
```diff
@@ -0,0 +1,80 @@
+
+class ReportGeneratorImplTest {
+    private static final String HEADER = "fruit,quantity";
+    private static final String LINE_SEPARATOR = System.lineSeparator();
```
don't create constant for static method

Mwxyz2022 committed:
fix field

liliia-ponomarenko reviewed (approved):
Good job!
